File: z/rainbow_sort.cc
class Solution {
 public:
  vector<int> rainbowSort(vector<int> array) {
    // write your solution here
    if (array.size() <= 1) {
      return array;
    }
    int start = 0, index = 0, end = array.size() - 1;
    while (index <= end) {
      if (array[index] == -1) {
        swap(array[start++], array[index++]);
      } else if (array[index] == 0) {
        ++index;
      } else if (array[index] == 1) {
        swap(array[index], array[end--]);
      }
    }
    return array;
  }
};

File: h/RainbowSort.java
public class Solution {
  public int[] rainbowSort(int[] array) {
    if (array.length == 0) {
      return array;
    }
    int i = 0;
    int j = 0;
    int k = array.length - 1;

    while (j <= k) {
      if (array[j] == -1) {
        swap(array, i++, j++);
      } else if (array[j] == 0) {
        j++;
      } else if (array[j] == 1) {
        swap(array, j, k--);
      }
    }
    return array;
  }

  private void swap(int[] arr, int index1, int index2) {
    if (index1 == index2)
      return;
    arr[index1] = arr[index1] + arr[index2];
    arr[index2] = arr[index1] - arr[index2];
    arr[index1] = arr[index1] - arr[index2];
  }
}
